src/AggregateFunctions/AggregateFunctionGroupArraySorted.cpp

```diff
@@ -0,0 +1,394 @@
+        }
+    }
+};
+
+template <typename T>
+using GroupArraySortedDataHeap = GroupArraySortedData<T, GroupArraySortedStrategy::heap>;
+
+template <typename T>
+using GroupArraySortedDataSort = GroupArraySortedData<T, GroupArraySortedStrategy::sort>;
+
+constexpr UInt64 aggregate_function_group_array_sorted_max_element_size = 0xFFFFFF;
+
+template <typename Data, typename T>
+class GroupArraySorted final
+    : public IAggregateFunctionDataHelper<Data, GroupArraySorted<Data, T>>
+{
+public:
+    explicit GroupArraySorted(
+        const DataTypePtr & data_type_, const Array & parameters_, UInt64 max_elements_)
+        : IAggregateFunctionDataHelper<Data, GroupArraySorted<Data, T>>(
+            {data_type_}, parameters_, std::make_shared<DataTypeArray>(data_type_))
+        , max_elements(max_elements_)
+        , serialization(data_type_->getDefaultSerialization())
+    {
+        if (max_elements > aggregate_function_group_array_sorted_max_element_size)
+            throw Exception(ErrorCodes::BAD_ARGUMENTS,
+                "Too large limit parameter for groupArraySorted aggregate function, it should not exceed {}",
+                aggregate_function_group_array_sorted_max_element_size);
+    }
+
+    String getName() const override { return "groupArraySorted"; }
+
+    void add(AggregateDataPtr __restrict place, const IColumn ** columns, size_t row_num, Arena * arena) const override
+    {
+        if constexpr (std::is_same_v<T, Field>)
+        {
+            auto row_value = (*columns[0])[row_num];
+            this->data(place).addElement(std::move(row_value), max_elements, arena);
+        }
+        else
+        {
+            auto row_value = assert_cast<const ColumnVector<T> &>(*columns[0]).getData()[row_num];
+            this->data(place).addElement(std::move(row_value), max_elements, arena);
+        }
+    }
+
+    void merge(AggregateDataPtr __restrict place, ConstAggregateDataPtr rhs, Arena * arena) const override
+    {
+        auto & rhs_values = this->data(rhs).values;
+        for (auto rhs_element : rhs_values)
+            this->data(place).addElement(std::move(rhs_element), max_elements, arena);
+    }
+
+    void serialize(ConstAggregateDataPtr __restrict place, WriteBuffer & buf, std::optional<size_t> /* version */) const override
+    {
+        auto & values = this->data(place).values;
+        size_t size = values.size();
+        writeVarUInt(size, buf);
+
+        if constexpr (std::is_same_v<T, Field>)
+        {
+            for (const Field & element : values)
+            {
+                if (element.isNull())
+                {
+                    writeBinary(false, buf);
+                }
+                else
+                {
+                    writeBinary(true, buf);
+                    serialization->serializeBinary(element, buf, {});
+                }
+            }
+        }
+        else
+        {
+            for (const auto & element : values)
```

Comment:
We should batch this. It's going to be [much faster](https://github.com/ClickHouse/ClickHouse/pull/41089)

Reply:
Any thoughts on this? It should help writing and reading states from MVs. I'd rather make it fast now before we accept it as changing it later might mean a breaking change.

Reply:
Updated. Also updated `groupArray`.